### ISSUE TITLE
Readonly-ness is lost in callback parameter

### ISSUE THREAD

**Issue by trusktr** [Microsoft/TypeScript#28994]
**TypeScript Version:**  3.3.0-dev.20181208

**Search Terms:**

**Code**

```ts
type Events = {
    SOME_EVENT: number
    OTHER_EVENT: string
    ANOTHER_EVENT: undefined
    READONLY_EVENT: ReadonlyArray<number>
}

type EventArgs<EventTypes, K extends keyof EventTypes> = EventTypes[K] extends undefined ?[]:[EventTypes[K]]
interface EventEmitter<EventTypes> {
    on<K extends keyof EventTypes>(s: K, listener: (...v: EventArgs<EventTypes,K>) => void);
    emit<K extends keyof EventTypes>(s: K, ...v: EventArgs<EventTypes,K>);
}

declare const emitter: EventEmitter<Events>;

emitter.on('SOME_EVENT', (payload) => testNumber(payload));
emitter.on('OTHER_EVENT', (payload) => testString(payload));
emitter.on('ANOTHER_EVENT', () => {});
emitter.on('READONLY_EVENT', (a: Array<number>) => {}); // SHOULD BE ERROR!

emitter.emit('SOME_EVENT', 42)
emitter.emit('OTHER_EVENT', 'foo')
emitter.emit('ANOTHER_EVENT')
emitter.emit('READONLY_EVENT', Object.freeze([1,2,3]))

declare function testNumber(value: number)
declare function testString(value: string)
```

**Expected behavior:**

It should complain that the signature of the callback being passed into the `.on` call, on the line labeled with `SHOULD BE ERROR`, is wrong.

**Actual behavior:**

There's no error.

**Playground Link:**

[playground link](http://www.typescriptlang.org/play/#src=type%20Events%20%3D%20%7B%0A%20%20%20%20SOME_EVENT%3A%20number%0A%20%20%20%20OTHER_EVENT%3A%20string%0A%20%20%20%20ANOTHER_EVENT%3A%20undefined%0A%20%20%20%20READONLY_EVENT%3A%20ReadonlyArray%3Cnumber%3E%0A%7D%0A%0Atype%20EventArgs%3CEventTypes%2C%20K%20extends%20keyof%20EventTypes%3E%20%3D%20EventTypes%5BK%5D%20extends%20undefined%20%3F%5B%5D%3A%5BEventTypes%5BK%5D%5D%0Ainterface%20EventEmitter%3CEventTypes%3E%20%7B%0A%20%20%20%20on%3CK%20extends%20keyof%20EventTypes%3E(s%3A%20K%2C%20listener%3A%20(...v%3A%20EventArgs%3CEventTypes%2CK%3E)%20%3D%3E%20void)%3B%0A%20%20%20%20emit%3CK%20extends%20keyof%20EventTypes%3E(s%3A%20K%2C%20...v%3A%20EventArgs%3CEventTypes%2CK%3E)%3B%0A%7D%0A%0Adeclare%20const%20emitter%3A%20EventEmitter%3CEvents%3E%3B%0A%0Aemitter.on('SOME_EVENT'%2C%20(payload)%20%3D%3E%20testNumber(payload))%3B%0Aemitter.on('OTHER_EVENT'%2C%20(payload)%20%3D%3E%20testString(payload))%3B%0Aemitter.on('ANOTHER_EVENT'%2C%20()%20%3D%3E%20%7B%7D)%3B%0Aemitter.on('READONLY_EVENT'%2C%20(a%3A%20Array%3Cnumber%3E)%20%3D%3E%20%7B%7D)%3B%20%2F%2F%20SHOULD%20BE%20ERROR!%0A%0Aemitter.emit('SOME_EVENT'%2C%2042)%0Aemitter.emit('OTHER_EVENT'%2C%20'foo')%0Aemitter.emit('ANOTHER_EVENT')%0Aemitter.emit('READONLY_EVENT'%2C%20Object.freeze(%5B1%2C2%2C3%5D))%0A%0Adeclare%20function%20testNumber(value%3A%20number)%0Adeclare%20function%20testString(value%3A%20string)%0A)

**Related Issues:**

**Comment by weswigham** [comment]
It is flagged as an error with `strictFunctionTypes` (part of `strict`) on.

**Comment by weswigham** [comment]
`strictFunctionTypes` enables strict checking of argument variance. You don't "lose" anything - you just get stricter typechecking.
